FluxPulseScopeAnalysis improvements

stephlazar opened the pull request:
Changes proposed in this pull request:
- adds dosctring
- renames freq/delay_ranges_exclude to freq/delay_ranges_remove and fixes bug where only one range was being removed even when the used provided more ranges
- adds freq/delay_ranges_for_fit parameters to specify freq/delay ranges that should be included in the fit (opposite to rectangles_to_exclude).
- adds errorbars and marker to fitted frequency line and plot
- fixes bug where the sigma parameter in the fits was set to not be varied

With delay_ranges_to_fit specified:
![FluxPulseScope_qb5_20200907_155000 - Copy](https://user-images.githubusercontent.com/14844492/95296651-6d22e300-0879-11eb-9bb4-24fe569e3087.png)
![FluxPulseScope_qb5_20200907_155000 - Copy (2)](https://user-images.githubusercontent.com/14844492/95296660-714f0080-0879-11eb-97ed-f91d860eb235.png)

With delay_ranges_to_fit and freq_ranges_to_fit specified:
![FluxPulseScope_qb5_20200907_155000](https://user-images.githubusercontent.com/14844492/95296698-83c93a00-0879-11eb-8ad4-333755b1f021.png)

Commit:
analysis_V2, FluxPulseScopeAnalysis: improvements + docstring

- adds dosctring
- renames freq/delay_ranges_exclude to freq/delay_ranges_remove and fixes bug where only one range was being removed even when the used provided more ranges
- adds freq/delay_ranges_for_fit parameters to specify freq/delay ranges that should be included in the fit (opposite to rectangles_to_exclude).
- adds errorbars and marker to fitted frequency line and plot

Commit:
Merge branch 'qudev_master' into enh/fp_scope_analysis

Commit:
Merge branch 'qudev_master' into enh/fp_scope_analysis

Commit:
analysis_toolbox.py: adds zero_coord and one_coord to rotate_and_normalize_data_1ch

Commit:
MultiQubit_TimeDomain_Analysis: minor cleanup of cal points handling (adds do_pca flag) + implements fixed_cal_points_rotation for TwoD data with 2 cal states (adds minor class docstring about how to do rotation)

Commit:
MultiQubit_TimeDomain_Analysis: fix bug (missing absolute value) in rotate_data_TwoD_same_fixed_cal_idxs

Commit:
MultiQubit_TimeDomain_Analysis: Adds option to specify plot_params for the projected data plots.

Commit:
FluxPulseScopeAnalysis: changes marker type to 'x'

Commit:
MultiQubit_TimeDomain_Analysis: store rotation_coordinates in proc_data_dict if fixed_cal_points_rotation==True

Commit:
MultiQubit_TimeDomain_Analysis: adds option to do row-by-row PCA (do_PCA==True) or column-by-column PCA (do_column_PCA==True) + sets meaningful data axis label if any PCA is done.

Commit:
MultiQuit_TimeDomain_Analysis: removes all the different PCA flags and replaces them with rotation_type; updates docstring to reflect this change

chellings committed:
bugfix MultiQubit_TimeDomain_Analysis: rotation type column_PCA was called incorrectly in a function call

chellings committed:
tda.FluxPulseScopeAnalysis: improved handling of sign_of_peaks
- allow passing None (= use default) for individual qubits while passing a value for others
- default to +1 if rotation type fixed_cal_points is used

chellings committed:
tda column_PCA: rotate data such that the background has smaller values
background is detected by comparing median to mean of the data

chellings committed:
FluxPulseScopeAnalysis: default sign_of_peaks to +1 if column_PCA is used
This change is due to the new background sign convention in column_PCA.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -423,8 +441,7 @@ def process_data(self):
         for qbn in self.data_to_fit:
             if isinstance(self.data_to_fit[qbn], (list, tuple)):
                 self.data_to_fit[qbn] = self.data_to_fit[qbn][0]
-        global_PCA = self.get_param_value('global_PCA', default_value=False)
```
if this flag does not work any more, it needs to be removed from the docstring and the code in other modules needs to be refactored

stephlazar replied:
It was removed from the docstring of MultiQubit_TimeDomain_Analysis (the name appears there because it can be set as the rotation_type parameter). 
I just did a search of the whole timedomain_analysis.py module and I couldn't find the variable global_PCA anywhere. Did you find an exaplicit bug somewhere?

stephlazar replied:
Done now :)

chellings committed:
bugfix tda.FluxPulseScopeAnalysis: do not crash if sign_of_peaks is missing for some qubit
Use the default value for this qubit in that case.

Commit:
Merge branch 'qudev_master' into enh/fp_scope_analysis

Commit:
Merge branch 'qudev_master' into enh/fp_scope_analysis

Commit:
timedomain_analysis.py: raise deprecated parameter warning if global_PCA is passed + minor fix in MultiQubit_TimeDomain_Analysis docstring

Commit:
Updates use of global_PCA in measurement functions/classes to use rotation_type instead.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -878,7 +964,7 @@ def prepare_plots(self):
 
         if self.get_param_value('plot_raw_data', default_value=True):
-            # self.cal_states_rotations is not None:
+            # self.rotate == True:
```
you replace a comment with another comment. do we want that?

stephlazar replied:
removed now

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -980,8 +1066,10 @@ def prepare_projected_data_plot(
         if data_axis_label == '':
-            data_axis_label = '{} state population'.format(
-                self.get_latex_prob_label(self.data_to_fit[qb_name]))
+            data_axis_label = 'Rotated signal (arb.)' if \
```
should we write "strongest principal component (arb.)" ?

stephlazar replied:
Makes sense, I've renamed it now

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -7090,7 +7214,7 @@ def process_data(self):
                               slice_idx=None, data_slice=None):
         if slice_idx is None:
-            raise ValueError('"since_idx" cannot both be None. It is used '
+            raise ValueError('"since_idx" cannot be None. It is used '
```
shouldn't this be "slice_idx" ?

stephlazar replied:
Eagle eye :D Thanks! Fixed now

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -7101,7 +7225,7 @@ def prepare_fitting_slice(self, freqs, qbn, mu_guess,
         GaussianModel.set_param_hint('sigma',
                                      value=self.sigma_guess[qbn],
-                                     vary=False)
+                                     vary=True)
```
should we give the user the option to decide whether sigma should be fitted or fixed (to a user-provided value)? Maybe fitting it is less robust in some instances.

stephlazar replied:
I've fixed this by implementing a general way for the user to update guess pars. See the new method `set_user_guess_pars` in `BaseDataAnalysis`. Let me know what you think!

chellings replied:
I think the concept makes sense. I haven't done tests now, but if you say it works as intended, I am fine with it. Thank you!

chellings commented on pycqed/measurement/calibration/single_qubit_gates.py:
```diff
@@ -368,7 +370,7 @@ def run_analysis(self, **kw):
         self.analysis = tda.FluxPulseScopeAnalysis(
             qb_names=self.meas_obj_names,
-            options_dict=dict(TwoD=True, global_PCA=True,))
+            options_dict=dict(TwoD=True, rotation_type='global_PCA'))
```
should be 'fixed_cal_points' is cal states are used (and global_PCA otherwise). or maybe allow the user to pass a choice via analysis_kwargs (defaulting to what I have suggested in the previous sentence)?

stephlazar replied:
Fixed now

chellings reviewed (changes requested):
Thanks for all the improvements. I have used most of this on my test branch, and it works well. It might still make sense to include the following into this PR:
- include a cleaned-up version of the last two commits on https://github.com/QudevETH/PycQED_py3/commits/devel/fp_scope_analysis (5ade4a35a291d4f36349db88d87b02f6cb7737d5 and 5d5a7228bd1ac655ae8a4da266e69422a15b737e) as discussed recently (after figuring out how this can maybe be combined with Michael's data_mostly_g and deciding whether middle or median makes more sense)
- it would be nice to have an option to remove the fitted frequencies of g-level cal points from the plot (or not run fitting for them at all, an instead just set the fitted frequencies to nan). maybe we can discuss that idea.
- plus see the further reviewer comments

Commit:
Merge branch 'qudev_master' into enh/fp_scope_analysis

Commit:
timedomain_analysis.py: changes default pca y-axis label + minor cleanup

Commit:
ana_v2: option to let user update the guess_pars of a fit

- adds set_user_guess_pars to BaseDataAnalysis
- calls set_user_guess_pars in FluxPulseScopeAnalysis, RabiAnalysis, RamseyAnalysis, T1Analysis, QScaleAnalysis

Commit:
FluxPulseScope measurement class: updates run_analysis

Commit:
Merge branch 'qudev_master' into devel/fp_scope_analysis

Commit:
Analysis_v2: data_mostly_g cleanup

- renames a_tools.set_background_sign to set_majority_sign
- removes data_mostly_g functionality from a_tools.rotate_and_normalize_data_no_cal_points
- checks data_mostly_g flag for all types of PCA rotations in MultiQubit_TimeDomain_Analysis

Commit:
FluxPulseScopeAnalysis: adds flag fit_first_cal_state

Commit:
FluxPulseScopeAnalysis: recalculated cal_point sweep points when delays are removed

Commit:
MultiQubit_TimeDomain_Analysis: make the user-specified parameter data_mostly_g a dict instead of bool

chellings committed:
Merge branch 'enh/fp_scope_analysis' into devel/fp_scope_analysis

chellings committed:
FluxAmplitudeSweepAnalysis: set default values for rotation_type and data_mostly_g

chellings committed:
FluxPulseScopeAnalysis: set default values for rotation_type and data_mostly_g

chellings committed:
MultiQubit_TimeDomain_Analysis: enable data_mostly_g for column_PCA and PCA

chellings reviewed (approved):
I think we could merge now. Thanks for all your work in this context! Maybe you want to have a quick look at my three additional commits before doing the merge. We should announce in the pycqed channel that the way to activate global_PCA has been changed (people might have the old version in some notebooks).

stephlazar commented:
Shouldn't we add here the last 4 commits from devel/fp_scope_analysis? 
(be98bb59d801e0f7ff59cc0c835f996925836719,
 8038ee64cb733b06664b718c26e90637967b8f76,
 11f03451107ec7814b99012b843b2e6fc54f645e,
 71fc6752e885a18debb25b7d2b0b492c62ff6356)

chellings committed:
Revert "MultiQubit_TimeDomain_Analysis: enable data_mostly_g for column_PCA and PCA"

This reverts commit 56f50a58717f3d811f7e9082c13039e38efb2fd4.

chellings committed:
Merge branch 'devel/fp_scope_analysis' into enh/fp_scope_analysis

chellings committed:
Revert "MultiQubit_TimeDomain_Analysis: make the user-specified parameter data_mostly_g a dict instead of bool"

This reverts commit 71fc6752e885a18debb25b7d2b0b492c62ff6356.

chellings committed:
tda: changed behavior of rotations without cal points and data_mostly_g
Now, we always call set_majority_sign when rotating without cal points.
Setting data_mostly_g to False now means that the data is expected to be mostly in e level.
The detault of data_mostly_g was changed to True.

chellings committed:
tda.FluxPulseScopeAnalysis: extended automatic choice of sign_of_peaks to all types of PCA

chellings committed:
tda.FluxPulseScopeAnalysis: fixed clabel for cases where PCA is used

chellings commented:
Sorry for the mess, I had mixed up the branches. It should now be cleaned up including the additional changes that we have discussed. Before you merge, please check whether it makes sense to you. Thank you!

chellings committed:
Merge branch 'qudev_master' into enh/fp_scope_analysis

chellings committed:
bugfix FluxPulseScopeAnalysis: properly find 2D sweep parameter

chellings committed:
tda.FluxAmplitudeSweepAnalysis: use TwoD by default

chellings committed:
tda.FluxPulseScopeAnalysis: use TwoD by default

chellings committed:
FluxPulseScopeAnalysis: save processed delays_for_fit in analysis_params_dict

chellings committed:
bugfix tda.FluxPulseScopeAnalysis: extract fit keys for correct qubit (to enable parallel fp scope)

stephlazar commented:
Looks good now! Thanks for the cleanup!
